src/libxtp/eigencuda.cc

```diff
@@ -0,0 +1,197 @@
+/*
+ *            Copyright 2009-2019 The VOTCA Development Team
+ *                       (http://www.votca.org)
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *              http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <votca/xtp/eigencuda.h>
+
+namespace votca {
+namespace xtp {
+
+template <typename T>
+EigenCuda<T>::~EigenCuda() {
+  cublasDestroy(_handle);
+  for (auto &p : _allocated) this->fun_free(p.second);
+}
+
+template <typename T>
+void EigenCuda<T>::fun_alloc(T **x, std::size_t n) const {
+  // Allocate memory in the device
+  (_pinned) ? cudaMallocHost(x, n) : cudaMalloc(x, n);
+}
+
+template <typename T>
+void EigenCuda<T>::fun_free(T *x) const {
+  // Deallocate memory from the device
+  (_pinned) ? cudaFreeHost(x) : cudaFree(x);
+};
+
+template <typename T>
+void EigenCuda<T>::free_matrix(unsigned id) {
+  // Free Array with id from the device
+  fun_free(_allocated.at(id));
+  _allocated.erase(id);
+}
+
+template <typename T>
+unsigned EigenCuda<T>::initialize_Matrix(Mat<T> &A, bool copy_to_device) {
+  // Copy two matrices to the device
```

Comment:
two matrices? What dies the bool do?

Reply:
Copy a single matrix (typo).
the bool is there because in some cases you just want to allocate space for arrays but do not want to copy then to device